source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc

```diff
@@ -123,15 +124,137 @@ AWSLambdaFilter::decodeHeaders(Http::RequestHeaderMap &headers,
+    // return response as is if not configured for alb mode
+    return Http::FilterDataStatus::Continue;
+  }
+
+  // support for streaming and http2 for future proofing
+  encoder_callbacks_->addEncodedData(data, false);
+
+  if (!end_stream) {
+    // we need the entire response prior to parsing and unwrapping the json
+    return Http::FilterDataStatus::StopIterationAndBuffer;
+  }
+
+  // Now that the response is finished we know that the following is safe
+  // as the following options will only make the resulting buffer smaller.
+  const Buffer::Instance&  buff = *encoder_callbacks_->encodingBuffer();
+  encoder_callbacks_->modifyEncodingBuffer([this](Buffer::Instance& enc_buf) {
+    Buffer::OwnedImpl body;
+    parseResponseAsALB(*response_headers_, enc_buf, body);
+    enc_buf.drain(enc_buf.length());
+    enc_buf.move(body);
+  });
+  response_headers_->setContentLength(buff.length());
+  return Http::FilterDataStatus::Continue;
+}
+
+Http::FilterTrailersStatus 
+AWSLambdaFilter::encodeTrailers(Http::ResponseTrailerMap &) {
+
+  if (functionOnRoute() == nullptr || !functionOnRoute()->unwrapAsAlb()){
+   return Http::FilterTrailersStatus::Continue;
+  }
+  // Future proof against alb http2 support and finalize the data transform
+  const Buffer::Instance&  buff = *encoder_callbacks_->encodingBuffer();
+  encoder_callbacks_->modifyEncodingBuffer([this](Buffer::Instance& enc_buf) {
+    Buffer::OwnedImpl body;
+    parseResponseAsALB(*response_headers_, enc_buf, body);
+    enc_buf.drain(enc_buf.length());
+    enc_buf.move(body);
+  });
+  response_headers_->setContentLength(buff.length());
+  return Http::FilterTrailersStatus::Continue;
+}
+
+
+void AWSLambdaFilter::parseResponseAsALB(Http::ResponseHeaderMap& headers, 
+                    const Buffer::Instance& json_buf, Buffer::Instance& body) {
+
+  ProtobufWkt::Struct alb_response;                
+  try {
+    MessageUtil::loadFromJson(json_buf.toString(), alb_response);
+  } catch (EnvoyException& ex) {
+    ENVOY_LOG(warn, "{}: alb_unwrap set but did not recieve a json payload", 
+                                                   functionOnRoute()->path());
+    headers.setStatus(static_cast<int>(Http::Code::InternalServerError));
+    return;
+  }
+  
+  bool is_bad_format = false;
+  const auto& flds = alb_response.fields();
+  if (flds.contains("body")) {
+    auto rawBody = flds.at("body").string_value();
+    if (flds.contains("isBase64Encoded")){
+      is_bad_format = is_bad_format ||
+                             !flds.at("isBase64Encoded").has_bool_value();
+      if (flds.at("isBase64Encoded").bool_value()){
+        rawBody = Base64::decode(rawBody);
+      }
+    }
+    body.add(rawBody);
+  }
+
+  if (flds.contains("statusCode")){
+    headers.setStatus(flds.at("statusCode").number_value());
+    is_bad_format= is_bad_format || !flds.at("statusCode").has_number_value();
+  }
+  if (flds.contains("headers")){
+    auto hds = flds.at("headers").struct_value();
```

Comment:
use `const auto&` to avoid unneeded copying on the datapath (here and below)

Reply:
Ive changed this for all the portions that dont get modified. Which leaves body as we may overwrite it in the case of base64